Source/Error.swift

```diff
@@ -24,41 +24,540 @@
+/// `AFError` is the error type returned by the Alamofire framework. It encompasses a few
+/// different types of errors, each with their own associated reasons.
+///
+/// `.multipartEncodingFailed` errors are returned when some step in the multipart encoding process fails.
+///
+/// `.responseValidationFailed` errors are returned when a `validate()` call fails.
+///
+/// `.responseSerializationFailed` errors are returned when a response serializer encounters an error in the serialization process.
+public enum AFError: Error {
+
+    public enum MultipartEncodingFailureReason {
+        case bodyPartURLInvalid(at: URL)
+        case bodyPartFilenameInvalid(in: URL)
+        case bodyPartFileNotReachable(at: URL)
+        case bodyPartFileNotReachableWithError(url: URL, error: Error)
+        case bodyPartFileIsDirectory(at: URL)
+        case bodyPartFileSizeNotAvailable(at: URL)
+        case bodyPartFileSizeQueryFailedWithError(url: URL, error: Error)
+        case bodyPartInputStreamCreationFailed(for: URL)
+
+        case outputStreamCreationFailed(for: URL)
+        case outputStreamFileAlreadyExists(at: URL)
+        case outputStreamURLInvalid(url: URL)
+        case outputStreamWriteFailed(error: Error)
+
+        case inputStreamReadFailed(error: Error)
+    }
+
+    public enum ValidationFailureReason {
+        case missingContentType(acceptable: [String])
+        case unacceptableContentType(acceptable: [String], response: String)
+        case unacceptableStatusCode(code: Int)
+    }
+
+    public enum SerializationFailureReason {
+        case inputDataNil
+        case inputDataNilOrZeroLength
+        case stringSerializationFailed(encoding: String.Encoding)
+        case jsonSerializationFailed(error: Error)
+        case propertyListSerializationFailed(error: Error)
+    }
+
+    case multipartEncodingFailed(reason: MultipartEncodingFailureReason)
+    case responseValidationFailed(reason: ValidationFailureReason)
+    case responseSerializationFailed(reason: SerializationFailureReason)
+
+}
+
+// MARK: Error Booleans
+
+public extension AFError {
+
+    /// Returns whether the AFError is a multipart encoding error.
+    /// When true, the `url` and `underlyingError` properties will contain the associated values.
+    public var isMultipartEncodingError: Bool {
+        if case .multipartEncodingFailed = self { return true }
+        return false
+    }
+
+    /// Returns whether the `AFError` is a response validation error.
+    /// When true, the `acceptableContentTypes`, `responseContentType`, and `responseCode` properties will contain the associated values.
+    public var isResponseValidationError: Bool {
+        if case .responseValidationFailed = self { return true }
+        return false
+    }
+
+    /// Returns whether the `AFError` is a response serialization error.
+    /// When true, the `failedStringEncoding` and `underlyingError` properties will contain the associated values.
```

Comment:
Why'd you start putting different sentences on different lines? Everywhere else in AF we would do the following:

```
/// Returns whether the `AFError` is a response serialization error. When true, the `failedStringEncoding` and 
/// `underlyingError` properties will contain the associated values.
```

They just go out the 120 character limit then pull down to the next line. Reasoning for this is to make it easier to review so you don't have to scroll left-to-right throughout the diff in GitHub.

Same comment applies to the two docstrings above this one as well.